word_embeddings.py

```diff
@@ -0,0 +1,318 @@
+    Attributes:
+        fnames: list of file names that contains crashes data.
+        corpus: The pre-processed data.
+        model: The trained model.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, path):
+        self.fnames = path
+        self.corpus = self._read_corpus()
+        self.model = self._train_model()
+
+    @abstractmethod
+    def _read_corpus(self):
+        pass
+
+    def _train_model(self):
+        pass
+
+    @staticmethod
+    def _wmdistance(model, words1, words2, all_distances):
+        # Code modified from https://github.com/RaRe-Technologies/gensim/blob/4f0e2ae/gensim/models/keyedvectors.py#L339
+        dictionary = gensim.corpora.Dictionary(documents=[words1, words2])
+        vocab_len = len(dictionary)
+
+        # create bag of words from document
+        def create_bow(doc):
+            norm_bow = np.zeros(vocab_len, dtype=np.double)
+            bow = dictionary.doc2bow(doc)
+
+            for idx, count in bow:
+                norm_bow[idx] = count / float(len(doc))
+
+            return norm_bow
+
+        bow1 = create_bow(words1)
+        bow2 = create_bow(words2)
+
+        docset = set(words2)
+
+        def create_distance_matrix(model, dictionary, docset, all_distances):
+            distances = np.zeros((len(dictionary), len(dictionary)), dtype=np.double)
+            for j, w in dictionary.items():
+                if w in docset:
+                    distances[:all_distances.shape[1], j] = all_distances[model.wv.vocab[w].index].transpose()
+
+            return distances
+
+        distances = create_distance_matrix(model, dictionary, docset, all_distances)
+
+        assert sum(distances) != 0
+        return emd(bow1, bow2, distances)
+
+    def signature_similarity(self, paths, signature1, signature2):
+        model = self.model
+        model.init_sims(replace=True)
+        traces1 = crash_similarity.get_stack_traces_for_signature(paths, signature1)
+        traces2 = crash_similarity.get_stack_traces_for_signature(paths, signature2)
+
+        similarities = []
+        already_processed = set()
+
+        for doc1 in traces1:
+            words1 = np.unique([word for word in crash_similarity.preprocess(doc1) if word in model]).tolist()
+            distances = np.array(1.0 - np.dot(model.wv.syn0norm, model.wv.syn0norm[
+                [model.wv.vocab[word].index for word in words1]].transpose()), dtype=np.double)
+
+            for doc2 in traces2:
+                words2 = [word for word in crash_similarity.preprocess(doc2) if word in model]
+
+                if words1 == words2 or frozenset([tuple(words1), tuple(words2)]) in already_processed:
+                    continue
+                already_processed.add(frozenset([tuple(words1), tuple(words2)]))
+
+                similarities.append((doc1, doc2, self.wmdistance(model, words1, words2, distances)))
+
+        return sorted(similarities, key=lambda v: v[2])
+
+
+class Word2Vec(EmbeddingAlgo):
```

Comment:
There's a lot of duplicated code between the `Word2Vec` and `Doc2Vec` classes. It should be possible to put some common things in utility functions or in the base class.
E.g. there could be a `read_traces` in the base class (or as a utility function) that does something like this:
```Python
def read_traces():
    already_selected = set()
    for line in utils.read_files(self.fnames):
        data = json.loads(line)
        proto_signature = data['proto_signature']

        if crash_similarity.should_skip(proto_signature):
            continue

        processed = crash_similarity.preprocess(proto_signature)

        if frozenset(processed) not in already_selected:
            yield elems
```

Then you can use it in `Word2Vec` and `Doc2Vec`, e.g. for `Word2Vec`:
```Python
def read_corpus():
    return [elem for elem in self.read_traces()]
```